packages/react-router-dom/index.tsx

```diff
@@ -403,40 +413,50 @@ export const NavLink = React.forwardRef<HTMLAnchorElement, NavLinkProps>(
     },
     ref
   ) {
-    let location = useLocation();
     let path = useResolvedPath(to);
+    let match = useMatch({ path: path.pathname, end, caseSensitive });
+
+    let routerState = React.useContext(UNSAFE_DataRouterStateContext);
+    let nextLocation = routerState?.navigation.location;
+    let nextPath = useResolvedPath(nextLocation || "");
+    let nextMatch = React.useMemo(
+      () =>
+        nextLocation
+          ? matchPath(
+              { path: path.pathname, end, caseSensitive },
+              nextPath.pathname
+            )
+          : null,
+      [nextLocation, path.pathname, caseSensitive, end, nextPath.pathname]
+    );
```

Comment:
Calculate whether this matches the "next" location in the same way `useMatch` does internally, just using `navigation.location` instead of `useLocation()`